### PR TITLE
feat(electric): Add server-side validation of null values incoming from Satellite clients 

### PR DESCRIPTION
**Pull request by alco**
Admitedly, we're relying on client-provided SatRelation messages to decide whether any given column is nullable or not. So it's not a good protection against malicious clients.

I wish we didn't have to pass nullability info via SatRelation message from the server to the client and back. But apparently that's what we have to do since we're not storing the history of schema changes for every individual client.

This has to be tagged as a feature because it changes the SatRelationColumn message in the protobuf definition.

This PR is based on the changes made in https://github.com/electric-sql/electric/pull/320. Related to VAX-813.

### CODE REVIEW

File: protocol/satellite.proto [hunk 1]
```diff
@@ -176,6 +176,7 @@ message SatRelationColumn {
     string name = 1;
     string type = 2;
     bool primaryKey = 3;
+    bool is_nullable = 4;
```
Comment (magnetised) [inline protocol/satellite.proto:179]: aren't we using camel case for the protocol??

**Review by magnetised — COMMENTED** [review]
that was a lot of work, thanks. was about to hit approve but shouldn't the protocol message field be `isNullable`? bikeshedding!

File: protocol/satellite.proto (hunk 1 above)
Comment (magnetised) [reply protocol/satellite.proto:179]: no, it seems we're not! my bad

**Comment by magnetised** [comment]
ignore my comment (😊) - i looked at the rest of that protobuf definitition and it seems my memory was at fault, not your code....

**Review by magnetised — APPROVED** [review]
sorry!

File: components/electric/lib/electric/postgres/replication.ex [hunk 2]
```diff
@@ -10,17 +10,19 @@ defmodule Electric.Postgres.Replication do
...
     ]
 
     @type name() :: Postgres.name()
 
     @type t() :: %__MODULE__{
-            name: name(),
-            type: atom(),
-            type_modifier: integer(),
-            part_of_identity?: boolean() | nil
+            name: name,
```
[the remark this answers is not in the document]
Comment (alco) [reply components/electric/lib/electric/postgres/replication.ex:21]: It was me editing those types by hand. Elixir's builtin types don't need parens and aren't automatically formatted with parens. I don't know how we got parens here in the first place. I've reverted this change to keep the style consistent across the whole file.

File: components/electric/lib/electric/satellite/serialization.ex [hunk 3]
```diff
@@ -389,18 +391,20 @@ defmodule Electric.Satellite.Serialization do
...
-      {_type, _val, true = _null?} ->
+  defp validate_values!(%SatOpRow{nulls_bitmask: bitmask, values: values} = row, columns) do
+    Enum.each(Enum.zip([columns, values, bitmask_to_boolean(bitmask)]), fn
+      {%{nullable?: true}, _val, true = _null?} ->
         :ok
 
-      {type, val, false = _null?} when is_binary(val) ->
-        Electric.Satellite.Validation.assert_type!(val, type)
+      {%{nullable?: false}, _val, true = _null?} ->
+        raise "protocol violation, null value for a not null column"
```
[the remark this answers is not in the document]
Comment (alco) [reply components/electric/lib/electric/satellite/serialization.ex]: I know, right? It's maybe possible to implement it without the roundtrip through Satellite by keeping additional state on the server for each client.

**Pull request merged**